include/modem/modem_trace.h

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+/**@file modem_trace.h
+ *
+ * @defgroup modem_trace nRF91 Modem traces
+ * @{
+ */
+#ifndef MODEM_TRACE_H__
+#define MODEM_TRACE_H__
+
+#include <stdint.h>
+
+/** @brief Initialize the modem trace module.
+ *
+ * Initializes the trace transport medium.
+ *
+ * @return Zero on success, non-zero otherwise.
+ */
+int modem_trace_init(void);
+
+/**@brief Trace mode
+ *
+ * The trace mode can be used to filter the traces.
+ */
+enum modem_trace_mode {
+	MODEM_TRACE_COREDUMP_ONLY = 1,  /**< Coredump only. */
+	MODEM_TRACE_ALL = 2,            /**< LTE, IP, GNSS, and coredump. */
+	MODEM_TRACE_IP_ONLY = 4,        /**< IP. */
+	MODEM_TRACE_LTE_IP = 5,         /**< LTE and IP. */
+};
+
+/** @brief Start a trace session.
+ *
+ * This function sends AT command that requests the modem to start sending traces.
+ *
+ * @param trace_mode Trace mode
+ * @param duration   Trace duration in seconds. If set to 0, the trace session will
+ *                   continue until @ref modem_trace_stop is called or until
+ *                   the required size of max trace data (specified by the
+ *                   @p max_size parameter) is received.
+ * @param max_size   Maximum size (in bytes) of trace data that should be received.
+ *                   The tracing will be stopped after receiving @p max_size
+ *                   bytes. If set to 0, the trace session will continue until
+ *                   @ref modem_trace_stop is called or until the duration
+ *                   set via the @p duration parameter is reached.
+ *                   To ensure the integrity of the trace output, the
+ *                   modem_trace module will never skip a trace message. For
+ *                   this purpose, if it detects that a received trace won't fit
+ *                   in the maximum allowed size, it will stop the trace session
+ *                   without sending out that trace to the transport medium.
+ *
+ * @return Zero on success, non-zero otherwise.
+ */
+int modem_trace_start(enum modem_trace_mode trace_mode, uint16_t duration, uint32_t max_size);
+
+/** @brief Process modem trace data
+ *
+ * This function should only be called to process a trace received from the modem by the
+ * @ref nrf_modem_os_trace_put function. This function forwards the trace to the selected
```

Comment:
If this is intended for use only by a specific function, I wonder if it would be cleaner not to have it in the API and and use declaration with `extern` keyword in `nrf_modem_os.c`.

Up to you guys to decide which way to go but I'd always like to see public functions to be something that the user of the API should be interested in...

Reply:
It need not be called from nrf_modem_os_trace_put() alone. It can ,for example, be called from a thread/work queue that is handling the traces from the modem. No matter where this function is called, the input should be called with the trace buffer sent to the application via the `nrf_modem_os_trace_put()` function and not something random.